File: python/src/reality_capture/service/response.py
from typing import TypeVar, Generic, Optional
from reality_capture.service.error import DetailedErrorResponse


T = TypeVar("T")


class Response(tuple, Generic[T]):
    """
    A tuple containing a Service response.
    """

    status_code: int
    error: Optional[DetailedErrorResponse]
    value: Optional[T]

    def __new__(cls, status_code: int, error: Optional[DetailedErrorResponse], value: Optional[T]):
        self = tuple.__new__(cls, (status_code, error, value))
        self.value = value
        self.status_code = status_code
        self.error = error
        return self

    def get_response_status_code(self) -> int:
        """
        Return the HTTP response status_code

        Returns:
            int. The HTTP response status_code
        """
        return self.status_code

    def is_error(self) -> bool:
        """
        Checks whether the response is an error response.

        Returns:
            True if the response contains a valid error.
        """
        return self.error is not None

File: typescript_sdk/reality_capture/src/service/response.ts
import { DetailedErrorResponse } from './error';

export class Response<T> extends Array<any> {
    status_code: number;
    error?: DetailedErrorResponse | null;
    value?: T | null;

    /**
     * A tuple containing a Service response.
     * @param status_code HTTP Status Code returned by the service.
     * @param error Optional error if the request failed.
     * @param value Optional object if the request succeeded.
     */
    constructor(status_code: number, error?: DetailedErrorResponse | null, value?: T | null) {
        super();
        this.status_code = status_code;
        this.error = error ?? null;
        this.value = value ?? null;
        this.push(status_code, error ?? null, value ?? null);
    }

    /**
     * Return the HTTP response status_code
     * @returns int. The HTTP response status_code
     */
    getResponseStatusCode(): number {
        return this.status_code;
    }

    /**
     * Checks whether the response is an error response.
     * @returns True if the response contains a valid error.
     */
    isError(): boolean {
        return this.error !== null && this.error !== undefined;
    }
}